htlcswitch/link_test.go

```diff
@@ -0,0 +1,615 @@
+
+	var amount btcutil.Amount = 4 * btcutil.SatoshiPerBitcoin
+	errChan, invoice, err := n.makePayment(n.carolServer, amount)
+	if err != nil {
+		t.Fatalf("can't make the payment from carol to alice: %v", err)
+	}
+
+	// Wait for:
+	// * HTLC add request to be sent to from Alice to Bob.
+	// * Alice<->Bob commitment states to be updated.
+	// * Bob trying to add HTLC add request in Bob<->Carol channel.
+	// * Cancel HTLC request to be sent back from Bob to Alice.
+	// * user notification to be sent.
+	select {
+	case err := <-errChan:
+		if err == nil {
+			t.Fatal("error wasn't received")
+		}
+		break
+	case <-time.After(time.Second):
+		t.Fatal("error was no settled in time")
+	}
+
+	// Wait for Alice to receive the revocation.
+	time.Sleep(100 * time.Millisecond)
+
+	// Check that alice invoice wasn't settled and bandwidth of htlc
+	// links hasn't been changed.
+	if invoice.Terms.Settled {
+		t.Fatal("alice invoice was settled")
+	}
+
+	if n.aliceChannelLink.Bandwidth() != aliceBandwidthBefore {
+		t.Fatal("the bandwidth of alice channel link which handles " +
+			"alice->bob channel should be the same")
+	}
+
+	if n.firstBobChannelLink.Bandwidth() != firstBobBandwidthBefore {
+		t.Fatal("the bandwidth of bob channel link which handles " +
+			"alice->bob channel should be the same")
+	}
+
+	if n.secondBobChannelLink.Bandwidth() != secondBobBandwidthBefore {
+		t.Fatal("the bandwidth of bob channel link which handles " +
+			"bob->carol channel should be the same")
+	}
+
+	if n.carolChannelLink.Bandwidth() != carolBandwidthBefore {
+		t.Fatal("the bandwidth of carol channel link which handles " +
+			"bob->carol channel should be the same")
+	}
+
+}
+
+// TestMultiHopUnknownPaymentHash checks that we receive remote error from Alice if she
+// received not suitable payment hash for htlc.
+func TestChannelLink_MultiHopUnknownPaymentHash(t *testing.T) {
+	n := newThreeHopNetwork(t)
+	if err := n.start(); err != nil {
+		t.Fatalf("can't start three hop network: %v", err)
+	}
+	defer n.stop()
+
+	carolBandwidthBefore := n.carolChannelLink.Bandwidth()
+	firstBobBandwidthBefore := n.firstBobChannelLink.Bandwidth()
+	secondBobBandwidthBefore := n.secondBobChannelLink.Bandwidth()
+	aliceBandwidthBefore := n.aliceChannelLink.Bandwidth()
+
+	var amount btcutil.Amount = btcutil.SatoshiPerBitcoin
+
+	// Generate route convert it to blob, and return next destination for
+	// htlc add request.
+	hop, blob, err := generateRoute([]Peer{
+		n.bobServer,
+		n.carolServer,
+	}...)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Generate payment: invoice and htlc.
+	invoice, htlc, err := generatePayment(amount, blob)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// We need to have wrong rhash for that reason we should change the
+	// preimage. Inverse first byte by xoring with 0xff.
+	invoice.Terms.PaymentPreimage[0] ^= byte(255)
+
+	// Check who is last in the route and add invoice to server registry.
+	n.carolServer.registry.AddInvoice(invoice)
```

Comment:
The error here should be checked, same goes for all other instances of `AddInvoice`.